compiler/noirc_evaluator/src/ssa/checks/check_for_underconstrained_values.rs

```diff
@@ -63,6 +92,355 @@ fn check_for_underconstrained_values_within_function(
+        // one of the results should be constrained
+        for (i, result_status) in self.results.iter().enumerate() {
+            match result_status {
+                // Skip checking already covered results
+                ResultStatus::Constrained => {
+                    continue;
+                }
+                ResultStatus::Unconstrained { descendants } => {
+                    if descendants.intersection(constrained_values).next().is_some() {
+                        results_involved.push(i);
+                    }
+                }
+            }
+        }
+
+        // Also, one of the argument descendants should be constrained
+        // (skipped if there were no arguments, or if an actual result and not a
+        // descendant has been constrained, e.g. against a constant)
+        if !results_involved.is_empty()
+            && (self.arguments.is_empty()
+                || self.root_results.intersection(constrained_values).next().is_some()
+                || self.arguments.intersection(constrained_values).next().is_some())
+        {
+            // Remember the partial constraint, clearing the sets
+            results_involved.iter().for_each(|i| self.results[*i] = ResultStatus::Constrained);
+        }
+    }
+}
+
+impl DependencyContext {
+    /// Build the dependency context of variable ValueIds, storing
+    /// information on value ids involved in unchecked Brillig calls
+    fn build(&mut self, function: &Function, all_functions: &BTreeMap<FunctionId, Function>) {
+        self.block_queue.push(function.entry_block());
+        while let Some(block) = self.block_queue.pop() {
+            if self.visited_blocks.contains(&block) {
+                continue;
+            }
+            self.visited_blocks.insert(block);
+            self.process_instructions(block, function, all_functions);
+        }
+    }
+
+    /// Go over the given block tracking Brillig calls and checking them against
+    /// following constraints
+    fn process_instructions(
+        &mut self,
+        block: BasicBlockId,
+        function: &Function,
+        all_functions: &BTreeMap<FunctionId, Function>,
+    ) {
+        trace!("processing instructions of block {} of function {}", block, function);
+
+        for instruction in function.dfg[block].instructions() {
+            // Optimization: there is no reason to do anything until a Brillig
+            // call is encountered
+            if self.tainted.is_empty() {
+                let mut brillig_encountered = false;
+                if let Instruction::Call { func: func_id, .. } = &function.dfg[*instruction] {
+                    if let Value::Function(callee) = &function.dfg[*func_id] {
+                        if all_functions[&callee].runtime().is_brillig() {
+                            brillig_encountered = true;
+                        }
+                    }
+                }
+                if !brillig_encountered {
+                    continue;
```

Comment:
So the rules to mark a Brillig call as constrained are that when we find a constraint, it should:
* be a descendant of one of the results of the call, and
  * be a descendant of one of the arguments of the call, or
  * constrain one of the immediate results of the call

The processing of instructions doesn't have a memory of constraints _before_ the Brillig call; only constraints that happen after the call are considered. 

I wonder what happens in an example such as this:

```rust
unconstrained fn times10(a: Field) -> Field { 
  a * 10
}
fn main(a: Field) {
  assert(a == 5);
  let b = times10(a);
  let c = 2 * b;
  assert(b == 100);
}
```

It feels like it shouldn't matter whether `a` gets constrained before or after the call to the Brillig function, but it looks like it might not realise because it would ignore that instruction.

Reply:
Ah, never mind, the second assertion should contain descendants of `a` _as well as_ `b`.